src/lib/Inference.hs

```diff
@@ -642,11 +647,58 @@ checkAnn ann = case ann of
   Nothing -> freshType TyKind
 
 checkUType :: UType -> UInferM Type
-checkUType ty = do
-  reduced <- typeReduceScoped $ withEffects Pure $ checkRho ty TyKind
+checkUType ty@(WithSrc ctx _) =
+  typeReduceScoped (withEffects Pure $ checkRho ty TyKind) >>=
+  typeReductionAsAtom ctx "Failed to reduce type annotation"
+
+-- Delayed unification task. When applied, tries to perform dictionary synthesis
+-- within the inner scope and normalize eafResult such that it contains no free
+-- vars from the eafInner scope (while it can refer to eafOuter scope).
+-- If the evaluation succeeds, the result is unified with eafExpected.
+data EvalAndUnif = EvalAndUnif
+  { eafOuter :: Scope
+  , eafInner :: Scope
+  , eafResult :: Atom
+  , eafScs :: SynthCandidates
+  , eafExpected :: Atom
+  , eafErr :: Err
+  }
+-- XXX: Those assume no shadowing between outer and local scopes
+instance HasVars EvalAndUnif where
+  freeVars EvalAndUnif{..} = (allFree `envDiff` eafInner) `envDiff` eafOuter
+    where allFree = freeVars eafResult <> foldMap freeVars eafInner <> foldMap freeVars eafOuter <> freeVars eafExpected
+-- XXX: This is incorrect if the subst is not scopeless
+instance Subst EvalAndUnif where
+  subst env EvalAndUnif{..} =
+    EvalAndUnif (subst env eafOuter) (subst env eafInner) (subst env eafResult)
+                (subst env eafScs) (subst env eafExpected) eafErr
+instance IsDelayedSolve EvalAndUnif where
+  performDelayedSolve EvalAndUnif{..} = do
+    innerNoDict <- zonk eafInner <&> fmap \case
+      (AtomBinderInfo ty (LetBound ann expr)) -> AtomBinderInfo ty $ LetBound ann expr'
+        where
+          comp = traverseExpr (traverseHoles synthDictTop) expr
+          expr' = case runSubstBuilderT eafOuter eafScs comp of
+            Failure  _   -> error $ pprint ty
+            Success (ans, _) -> ans
+      info -> info
+    case typeReduceLocals eafOuter innerNoDict (Atom eafResult) of
+      Right ans -> constrainEq ans eafExpected
+      Left  _   -> throwErr eafErr
+
+typeReductionAsAtom :: SrcPosCtx -> String -> Either (Scope, Atom) Atom -> UInferM Atom
+typeReductionAsAtom ctx msg reduced =
   case reduced of
-    Just ty' -> return $ ty'
-    Nothing  -> throw TypeErr $ "Can't reduce type expression: " ++ pprint ty
+    Right ty'         -> return $ ty'
+    Left (localScope, ty') -> do
+      outerScope <- getScope
+      env <- forM (localScope `envIntersect` freeVars ty') \(AtomBinderInfo ty _) ->
+        freshType ty <&> \case Var v -> v; _ -> error "Unreachable"
+      resTy <- substBuilder (env <&> Rename . varName) ty'
```

Comment:
So this is essentially renaming each of the free variables in `ty'` so that they instead refer to the new type variables created in the previous step?

Reply:
Correct. The idea here is that we evaluate as much as we can, but once we find variables in the local scope that can't be resolved to fully evaluated expressions, we replace them with inference variables and delay their full evaluation until the end of the inference process.